internal/engine/compiler/engine_cache.go

```diff
@@ -0,0 +1,183 @@
+package compiler
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+
+	"github.com/tetratelabs/wazero/internal/platform"
+	"github.com/tetratelabs/wazero/internal/u32"
+	"github.com/tetratelabs/wazero/internal/u64"
+	"github.com/tetratelabs/wazero/internal/wasm"
+)
+
+func (e *engine) deleteCodes(module *wasm.Module) {
+	e.mux.Lock()
+	defer e.mux.Unlock()
+	delete(e.codes, module.ID)
+
+	// Note: we do not call e.externCache.Delete, as the lifetime of
+	// the content is up to the implementation of extencache.ExternCache interface.
+}
+
+func (e *engine) addCodes(module *wasm.Module, codes []*code) (err error) {
+	e.addCodesToMemory(module, codes)
+	err = e.addCodesToExternCache(module, codes)
+	return
+}
+
+func (e *engine) getCodes(module *wasm.Module) (codes []*code, ok bool, err error) {
+	codes, ok = e.getCodesFromMemory(module)
+	if ok {
+		return
+	}
+	codes, ok, err = e.getCodesFromExternCache(module)
+	if ok {
+		e.addCodesToMemory(module, codes)
+	}
+	return
+}
+
+func (e *engine) addCodesToMemory(module *wasm.Module, codes []*code) {
+	e.mux.Lock()
+	defer e.mux.Unlock()
+	e.codes[module.ID] = codes
+}
+
+func (e *engine) getCodesFromMemory(module *wasm.Module) (codes []*code, ok bool) {
+	e.mux.RLock()
+	defer e.mux.RUnlock()
+	codes, ok = e.codes[module.ID]
+	return
+}
+
+func (e *engine) addCodesToExternCache(module *wasm.Module, codes []*code) (err error) {
+	if e.externCache == nil {
+		return
+	}
+	err = e.externCache.Add(module.ID, serializeCodes(codes))
+	return
+}
+
+func (e *engine) getCodesFromExternCache(module *wasm.Module) (codes []*code, hit bool, err error) {
+	if e.externCache == nil {
+		return
+	}
+
+	// Check if the entries exist in the external cache.
+	var cached io.ReadCloser
+	cached, hit, err = e.externCache.Get(module.ID)
+	if !hit || err != nil {
+		return
+	}
+	defer cached.Close()
+
+	// Otherwise, we hit the cache on external cache.
+	// We retrieve *code structures from `cached`.
+	var staleCache bool
+	codes, staleCache, err = deserializeCodes(cached)
+	if err != nil {
+		hit = false
+		return
+	} else if staleCache {
+		return nil, false, e.externCache.Delete(module.ID)
+	}
+
+	for i, c := range codes {
+		c.indexInModule = wasm.Index(i)
+		c.sourceModule = module
+	}
+	return
+}
+
+var (
+	wazeroMagic = "WAZERO"
+	// version must be synced with the tag of the wazero library.
+	version         = "1.0.0-dev"
```

Comment:
wondering if this is the right place to place version... maybe a dedicated pkg?

Reply:
yeah maybe like func-e we should show the flag that ends up used also by goreleaser to assign this.

```
// version is the string representation of globals.GlobalOpts
// We can't use debug.ReadBuildInfo because it doesn't get the last known version properly
// See https://github.com/golang/go/issues/37475
var version = "dev"
```

maybe by now we can use ReadBuildInfo instead?

Reply:
so wazero is library so we cannot use gorelease, but maybe `ReadBuildInfo` sounds promising

Reply:
oh yeah of course. 🤞 on ReadBuildInfo esp as we have a higher floor go version maybe can have some luck

Reply:
cool if wazero is used as the dependency, and inject the following code somewhere in wazero

```
	info, _ := debug.ReadBuildInfo()
	for _, dep := range info.Deps {
		if strings.Contains(dep.Path, "wazero") {
			fmt.Println(dep.Version)
		}
	}
```

then we get like `v0.0.0-20220818054749-076d3245e307`